frankenstein/chunk_store.go

```diff
@@ -0,0 +1,379 @@
+	incomingErrors := make(chan error)
+	hs := hours(from, through)
+	for _, hour := range hs {
+		for _, matcher := range matchers {
+			go func(hour model.Time, matcher *metric.LabelMatcher) {
+				if matcher.Type != metric.Equal {
+					incomingErrors <- fmt.Errorf("%s matcher not supported yet", matcher.Type)
+					return
+				}
+
+				// TODO escaping - this will break if label values contain the separator (:)
+				hashValue := fmt.Sprintf("%s:%s", hour, matcher.Name)
+				rangeMinValue := fmt.Sprintf("%s=%s:%s", matcher.Name, matcher.Value, minChunkID)
+				rangeMaxValue := fmt.Sprintf("%s=%s:%s", matcher.Name, matcher.Value, maxChunkID)
+
+				var resp *dynamodb.QueryOutput
+				err := timeRequest("Query", dynamoRequestDuration, func() error {
+					var err error
+					resp, err = c.dynamodb.Query(&dynamodb.QueryInput{
+						TableName: aws.String(c.tableName),
+						KeyConditions: map[string]*dynamodb.Condition{
+							hashKey: {
+								AttributeValueList: []*dynamodb.AttributeValue{
+									{S: aws.String(hashValue)},
+								},
+								ComparisonOperator: aws.String("EQ"),
+							},
+							rangeKey: {
+								AttributeValueList: []*dynamodb.AttributeValue{
+									{S: aws.String(rangeMinValue)},
+									{S: aws.String(rangeMaxValue)},
+								},
+								ComparisonOperator: aws.String("BETWEEN"),
+							},
+						},
+						ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+					})
+					return err
+				})
+				if resp.ConsumedCapacity != nil {
+					dynamoConsumedCapacity.WithLabelValues("Query").
+						Add(float64(*resp.ConsumedCapacity.CapacityUnits))
+				}
+				if err != nil {
+					log.Errorf("Error querying DynamoDB: %v", err)
+					incomingErrors <- err
+					return
+				}
+
+				chunkSet := []Chunk{}
+				for _, item := range resp.Items {
+					rangeValue := item[rangeKey].S
+					if rangeValue == nil {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					parts := strings.SplitN(*rangeValue, ":", 2)
+					if len(parts) != 2 {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					chunkValue := item[chunkKey].B
+					if rangeValue == nil {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					chunk := Chunk{
+						ID: parts[1],
+					}
+					if err := json.Unmarshal(chunkValue, &chunk); err != nil {
+						log.Errorf("Invalid item: %v", item)
+						incomingErrors <- err
+						return
+					}
+					chunkSet = append(chunkSet, chunk)
+				}
+
+				incomingChunkSets <- chunkSet
+			}(hour, matcher)
+		}
+	}
+
+	chunkSets := [][]Chunk{}
+	errors := []error{}
+	for i := 0; i < len(hs)*len(matchers); i++ {
+		select {
+		case chunkSet := <-incomingChunkSets:
+			chunkSets = append(chunkSets, chunkSet)
+		case err := <-incomingErrors:
+			errors = append(errors, err)
+		}
+	}
+	if len(errors) > 0 {
+		return nil, errors[0]
```

Comment:
For later: probably better to abort everything using contexts on the first error. Of course, the AWS stuff might not support aborting, but I guess we'd limit the parallelism on that anyways and not fire them off all at the same time, so that aborting will be worth it.